pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go

```diff
@@ -0,0 +1,282 @@
+		thread := c.bigThreadJunks[0][0]
+		remaining := c.bigThreadJunks[0][1:]
+		if len(remaining) >= c.threadsPerCore {
+			c.bigThreadJunks[0] = remaining
+		} else {
+			c.bigThreadJunks = c.bigThreadJunks[1:]
+			c.smallJunks = append(c.smallJunks, remaining...)
+		}
+		return &thread
+	}
+	return nil
+}
+
+func (c *cell) IsEmpty() bool {
+	return len(c.smallJunks) == 0 && len(c.bigThreadJunks) == 0
+}
+
+type cpuPool struct {
+	cells             []*cell
+	threadsPerCore    int
+	cores             int
+	allowCellCrossing bool
+}
+
+func NewStrictCPUPool(requestedToplogy *api.CPUTopology, nodeTopology *v1.Topology, cpuSet []int) VCPUPool {
+	return newCPUPool(requestedToplogy, nodeTopology, cpuSet, false)
+}
+
+func NewRelaxedCPUPool(requestedToplogy *api.CPUTopology, nodeTopology *v1.Topology, cpuSet []int) VCPUPool {
+	return newCPUPool(requestedToplogy, nodeTopology, cpuSet, true)
+}
+
+func newCPUPool(requestedToplogy *api.CPUTopology, nodeTopology *v1.Topology, cpuSet []int, allowCellCrossing bool) *cpuPool {
+	pool := &cpuPool{threadsPerCore: int(requestedToplogy.Threads), cores: int(requestedToplogy.Cores * requestedToplogy.Sockets), allowCellCrossing: allowCellCrossing}
+	cores := cpuJunksToCells(cpuSet, nodeTopology)
+
+	for _, coresOnCell := range cores {
+		c := cell{threadsPerCore: int(requestedToplogy.Threads)}
+		for j, core := range coresOnCell {
+			if len(core) >= c.threadsPerCore {
+				c.bigThreadJunks = append(c.bigThreadJunks, coresOnCell[j])
+			} else {
+				c.smallJunks = append(c.smallJunks, coresOnCell[j]...)
+			}
+		}
+		pool.cells = append(pool.cells, &c)
+	}
+	return pool
+}
+
+// cpuJunksToCells takes the allocated cpuset, determines which of the threads belongs to which cpu and which numa
+// cell and returns an aggregated view. The return map has the numa caell as index and contains for each cpu an array
+// of all allocated threads.
+func cpuJunksToCells(cpuSet []int, nodeTopology *v1.Topology) (cores [][][]uint32) {
+	threads := map[int]struct{}{}
+	visited := map[uint32]struct{}{}
+	cores = [][][]uint32{}
+	for _, cpu := range cpuSet {
+		threads[cpu] = struct{}{}
+	}
+	for _, cell := range nodeTopology.NumaCells {
+		var coresOnCell [][]uint32
+		for _, cpu := range cell.Cpus {
+			if _, exists := visited[cpu.Id]; exists {
+				continue
+			}
+			core := []uint32{}
+			if len(cpu.Siblings) == 0 {
+				visited[cpu.Id] = struct{}{}
+				if _, exists := threads[int(cpu.Id)]; exists {
+					core = append(core, cpu.Id)
+				}
+			} else {
+				for _, thread := range cpu.Siblings {
+					visited[thread] = struct{}{}
+					if _, exists := threads[int(thread)]; exists {
+						core = append(core, thread)
+					}
+				}
+			}
+			coresOnCell = append(coresOnCell, core)
+		}
+		cores = append(cores, coresOnCell)
+	}
+	return cores
+}
+
+func (p *cpuPool) FitCores() (cpuTune *api.CPUTune, err error) {
+	threads, remaining := p.fitCores(p.cores)
+
+	if remaining > 0 {
+		if p.allowCellCrossing {
```

Comment:
Is that differentiation just to cause different output and verify it in tests? Or do the two errors have functional impact?

Reply:
It is meant to log the correct reason and help users identify why the workload does not start. I think I have to slightly iterate over that because also in the first case there could just not be enough threads.

Reply:
Done.